src/Microsoft.VisualStudio.Threading/NonConcurrentSynchronizationContext.cs

```diff
@@ -0,0 +1,131 @@
+    /// Delegates will be invoked on the threadpool in the order they were posted with
+    /// <see cref="SynchronizationContext.Post(SendOrPostCallback, object)"/>.
+    /// No two delegates will ever be executed concurrently.
+    /// Note that if the delegate invokes an async method, the delegate formally ends
+    /// when the async method yields for the first time or returns, whichever comes first.
+    /// Once that delegate returns the next delegate can be executed.
+    /// </remarks>
+    public sealed class NonConcurrentSynchronizationContext : SynchronizationContext
+    {
+        /// <summary>
+        /// The queue of work to execute.
+        /// </summary>
+        private readonly AsyncQueue<(SendOrPostCallback, object?)> queue = new AsyncQueue<(SendOrPostCallback, object?)>();
+
+        /// <summary>
+        /// A value indicating whether to set this instance as <see cref="SynchronizationContext.Current" /> when invoking delegates.
+        /// </summary>
+        private readonly bool sticky;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NonConcurrentSynchronizationContext"/> class.
+        /// </summary>
+        /// <param name="sticky">
+        /// A value indicating whether to set this instance as <see cref="SynchronizationContext.Current" /> when invoking delegates.
+        /// This has the effect that async methods that are invoked on this <see cref="SynchronizationContext" />
+        /// will execute their continuations on this <see cref="SynchronizationContext" /> as well unless they use <see cref="Task.ConfigureAwait(bool)" /> with <c>false</c> as the argument.
+        /// </param>
+        public NonConcurrentSynchronizationContext(bool sticky)
+        {
+            this.sticky = sticky;
+
+            // Start the queue processor. It will handle all exceptions.
+            this.ProcessQueueAsync().Forget();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NonConcurrentSynchronizationContext"/> class
+        /// that is a copy of an existing instance.
+        /// </summary>
+        /// <param name="copyFrom">The instance to copy from.</param>
+        private NonConcurrentSynchronizationContext(NonConcurrentSynchronizationContext copyFrom)
+        {
+            // We do *not* kick off processing of the queue, since the original copy has done that and we don't want two logical threads processing the queue.
+            this.sticky = copyFrom.sticky;
+            this.queue = copyFrom.queue;
+        }
+
+        /// <summary>
+        /// Occurs when posted work throws an unhandled exception.
+        /// </summary>
+        public event EventHandler<Exception>? UnhandledException;
+
+        /// <inheritdoc />
+        public override void Post(SendOrPostCallback d, object? state) => this.queue.Enqueue((d, state));
+
+        /// <inheritdoc />
+        public override void Send(SendOrPostCallback d, object? state)
```

Comment:
Will this deadlock if the posted delegate itself tries to Send?

Reply:
> Will this deadlock if the sent delegate itself tries to Send?

With
> No two delegates will ever be executed concurrently.

If the implementation is true to that comment, `Send` must call the delegate directly if it is invoked when already on the right SynchronizationContext, otherwise you'd be either running two delegates simultanously or you would be deadlocking. I believe calling `Send` when already on the SC is normally part of the contract you have to support if you build a general purpose SC